### ISSUE TITLE
Commands history

### ISSUE THREAD

**Issue by lenormf** [gokcehan/lf#81]
Hi,

I'd like to be able to have the past few commands I ran inserted automatically in the promp upon hitting the up/down arrow keys. An option to control how many commands to keep in memory could probably tag along, I guess.

Thanks.

**Comment by gokcehan** [comment]
Just to be clear, you also mean going directly from normal mode to shell command mode with the prompt filled with the last command right? If so, as a side note, we have settled on using arrow keys for movements as default a while ago. I think we should use `<c-p>` and `<c-n>` as default instead. Anyway, I think we need this both as a normal mode key and a command mode key so you can move in the history while the command prompt is active. Also we should consider whether we should separate histories of waiting and non-waiting commands.

**Comment by lenormf** [comment]
I meant history about every command I've issued in the prompt (be it with `:`, `$` or `!`).

`<c-p>` and `<c-n>` are nice, but if we're going down that road, could we have a small subset of `readline` type shortcuts as well? I keep hitting `<c-g>` for nothing in the prompt, and I can't declare mappings within that particular scope.

**Comment by gokcehan** [comment]
@lenormf I have added `cmap` keyword a while ago which you can use to declare mappings for the prompt. It is already described in the documentation. Is that what you mean?

**Comment by lenormf** [comment]
Ah! I missed that.

**Comment by gokcehan** [comment]
I have pushed an implementation for this now. Two commands have been added `cmd-hist-prev` and `cmd-hist-next` and they can be mapped in both normal and command modes. Default bindings are `<c-p>` and `<c-n>` for both modes. I haven't added a history limit for now so everything is stored until exit.

I will keep this issue open for a while for feedback and testing.

**Comment by gokcehan** [comment]
Just to remind myself, this issue is still open since history is still not saved to files on exit and reloaded on startup. We may need to think about cases with multiple clients.

**Comment by gokcehan** [comment]
I have now added support for persistent history through the use of a history file. Clients read the file on startup and append their changes on quit. Multiple clients should append to this file in the order they quit and no history should be lost during this process.

History file is currently limited to 1000 commands. We can increase this or add an option for it though without history searching there is not much point in this for now. Closing this issue now.
